### ISSUE TITLE
Large objects are serialized incorrectly (Trac #2096)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2088]
It seems that at someone decided to store collection sizes in our `portable_binary_archive`s [http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/serialization/trunk/public/archive/portable_binary_archive.hpp#L73 as 32 bits]. While it might seem unlikely that we would ever need to serialize an array of 2^32^ objects, this can happen because `I3Frame::save` [http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/icetray/trunk/private/icetray/I3Frame.cxx#L566 serializes all frame objects to binary blobs], stored as `std::vector<char>`. Thus, any frame object whose serialized size exceeds 4GB has an incorrect size written (although its contents are still written out in full). The error becomes obvious on reading, when the implementation tries to read only the incorrectly reduced length, which usually causes the next read operation to fail (due to it unexpectedly occurring in the middle of the data of the very long object which hasn't been completely read) even before it can be discovered that the data which was read is incomplete (which would likely be detected by the checksumming). This is demonstrated fairly minimally by the following test case:

```text
#include <iostream>
#include <fstream>
#include <vector>
#include <icetray/serialization.h>

int main(){
	{ //write
		std::vector<char> v((1ULL<<32) + 100,'a');
		std::cout << "Writing vector with length " << v.size() << std::endl;
		std::ofstream outfile("test.dat",std::ios::out|std::ios::binary);
		icecube::archive::portable_binary_oarchive poa(outfile);
		poa << v;
	}
	{ //read
		std::vector<char> v;
		std::ifstream infile("test.dat",std::ios::in|std::ios::binary);
		icecube::archive::portable_binary_iarchive poa(infile);
		poa >> v;
		std::cout << "Read vector with length " << v.size() << std::endl;
	}
}
```

which can be compiled using 

```text
$CXX `icetray-config icetray`
```

The output is 

```text
Writing vector with length 4294967396
Read vector with length 100
```

which is clearly wrong. 

In order to fix this without breaking backwards compatibility, it seems that some fancy footwork will be required for `serialization::collection_size_type`, since we must continue to read 32 bit collection sizes (which are mostly correctly anyway). We should be able to replace the current, simple `sv_override` and `ld_override`, with something more clever which writes 32 bits by default, if the value being written fits in them, and otherwise writes a special value (2^32^-1?) and then records a 64 bit length, or similar. There is a minor issue that there is no length value which is technically `clean` for use to indicate overflow, since a file could validly contain any 32 bit pattern in that place, but it is genuinely unlikely that any file written so far contains an object of exactly the maximum allowed length. 

Old files containing large objects will be unrecoverable, as they are already corrupt, but it should be possible to support both old files with small objects (i.e. virtually all existing files) and be able to write and read large objects going forward. 

It should also be noted that we make size assumptions in several other places, such as the `save_override`s for string types, and all basic integer tags used by the serialization library itself: `version_type`, `class_id_type`, etc. Even if we could just make all of these things 64 bits, we clearly don't want to for size reasons. 

Migrated from https://code.icecube.wisc.edu/ticket/2096
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:14:49", 
    "description": "It seems that at someone decided to store collection sizes in our `portable_binary_archive`s [http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/serialization/trunk/public/archive/portable_binary_archive.hpp#L73 as 32 bits]. While it might seem unlikely that we would ever need to serialize an array of 2^32^ objects, this can happen because `I3Frame::save` [http://code.icecube.wisc.edu/projects/icecube/browser/IceTray/projects/icetray/trunk/private/icetray/I3Frame.cxx#L566 serializes all frame objects to binary blobs], stored as `std::vector<char>`. Thus, any frame object whose serialized size exceeds 4GB has an incorrect size written (although its contents are still written out in full). The error becomes obvious on reading, when the implementation tries to read only the incorrectly reduced length, which usually causes the next read operation to fail (due to it unexpectedly occurring in the middle of the data of the very long object which hasn't been completely read) even before it can be discovered that the data which was read is incomplete (which would likely be detected by the checksumming). This is demonstrated fairly minimally by the following test case:\n\n{{{\n#include <iostream>\n#include <fstream>\n#include <vector>\n#include <icetray/serialization.h>\n\nint main(){\n\t{ //write\n\t\tstd::vector<char> v((1ULL<<32) + 100,'a');\n\t\tstd::cout << \"Writing vector with length \" << v.size() << std::endl;\n\t\tstd::ofstream outfile(\"test.dat\",std::ios::out|std::ios::binary);\n\t\ticecube::archive::portable_binary_oarchive poa(outfile);\n\t\tpoa << v;\n\t}\n\t{ //read\n\t\tstd::vector<char> v;\n\t\tstd::ifstream infile(\"test.dat\",std::ios::in|std::ios::binary);\n\t\ticecube::archive::portable_binary_iarchive poa(infile);\n\t\tpoa >> v;\n\t\tstd::cout << \"Read vector with length \" << v.size() << std::endl;\n\t}\n}\n}}}\n\nwhich can be compiled using \n\n{{{\n$CXX `icetray-config icetray`\n}}}\n\nThe output is \n\n{{{\nWriting vector with length 4294967396\nRead vector with length 100\n}}}\n\nwhich is clearly wrong. \n\nIn order to fix this without breaking backwards compatibility, it seems that some fancy footwork will be required for `serialization::collection_size_type`, since we must continue to read 32 bit collection sizes (which are mostly correctly anyway). We should be able to replace the current, simple `sv_override` and `ld_override`, with something more clever which writes 32 bits by default, if the value being written fits in them, and otherwise writes a special value (2^32^-1?) and then records a 64 bit length, or similar. There is a minor issue that there is no length value which is technically `clean` for use to indicate overflow, since a file could validly contain any 32 bit pattern in that place, but it is genuinely unlikely that any file written so far contains an object of exactly the maximum allowed length. \n\nOld files containing large objects will be unrecoverable, as they are already corrupt, but it should be possible to support both old files with small objects (i.e. virtually all existing files) and be able to write and read large objects going forward. \n\nIt should also be noted that we make size assumptions in several other places, such as the `save_override`s for string types, and all basic integer tags used by the serialization library itself: `version_type`, `class_id_type`, etc. Even if we could just make all of these things 64 bits, we clearly don't want to for size reasons. ", 
    "reporter": "cweaver", 
    "cc": "cweaver", 
    "resolution": "fixed", 
    "_ts": "1550067289977126", 
    "component": "combo core", 
    "summary": "Large objects are serialized incorrectly", 
    "priority": "major", 
    "keywords": "", 
    "time": "2017-09-23T00:58:47", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170925T19:37:32: 
* david.schultz changed status from "new" to "accepted"
* david.schultz commented:

> `class_id_type`'s internal storage is `int16_t`, so that's fine.  Several others are similar, and won't be changed.

**Comment by icecube-trac** [comment]
Trac update at 20170927T17:11:20: 
* david.schultz changed _comment0 from "I think I've finished this.  Normally I would create a test, but I'm not sure the bots can stand to have a 5GB temp file, or more importantly a 5GB string." to "1506532307197706"
* david.schultz commented:

> I think I've finished this.  Normally I would create a test, but I'm not sure the bots can stand to have a 5GB temp file, or more importantly a 5GB string in memory.

**Comment by icecube-trac** [comment]
Trac update at 20170929T09:49:47: 
* david.schultz commented:

> In changeset 2977:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2977"
> Allow variable sizes for version_type, collection_size_type, and strings.  This allows for larger than 4GB frame objects. fixes #2088
> ```


* david.schultz changed resolution from "" to "fixed"
* david.schultz changed status from "accepted" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T23:29:57: nwhitehorn commented:

> In changeset 169996:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="169996"
> Allow variable sizes for version_type, collection_size_type, and strings.  This allows for larger than 4GB frame objects. fixes #2088
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:14:49: 
* olivas commented:

> Milestone October 2017 deleted


* olivas changed milestone from "October 2017" to ""
